core/trino-main/src/main/java/io/trino/operator/TableFunctionPartition.java

```diff
@@ -0,0 +1,405 @@
+            if (pagesIndex.isNull(markerChannel, mid)) {
+                end = mid;
+            }
+            else {
+                start = mid;
+            }
+        }
+        return end;
+    }
+
+    public record PassThroughColumnSpecification(boolean isPartitioningColumn, int inputChannel, int indexChannel)
+    {
+    }
+
+    private PassThroughColumnProvider createColumnProvider(PassThroughColumnSpecification specification)
+    {
+        if (specification.isPartitioningColumn()) {
+            return new PartitioningColumnProvider(pagesIndex.getSingleValueBlock(specification.inputChannel(), partitionStart));
+        }
+        return new NonPartitioningColumnProvider(specification.inputChannel(), specification.indexChannel());
+    }
+
+    private sealed interface PassThroughColumnProvider
+            permits PartitioningColumnProvider, NonPartitioningColumnProvider
+    {
+        Block getPassThroughColumn(Page page);
+    }
+
+    private record PartitioningColumnProvider(Block partitioningValue)
+            implements PassThroughColumnProvider
+    {
+        private PartitioningColumnProvider
+        {
+            requireNonNull(partitioningValue, "partitioningValue is null");
+        }
+
+        @Override
+        public Block getPassThroughColumn(Page page)
+        {
+            return RunLengthEncodedBlock.create(partitioningValue, page.getPositionCount());
+        }
+    }
+
+    private final class NonPartitioningColumnProvider
+            implements PassThroughColumnProvider
+    {
+        private final int inputChannel;
+        private final Type type;
+        private final int indexChannel;
+
+        public NonPartitioningColumnProvider(int inputChannel, int indexChannel)
+        {
+            this.inputChannel = inputChannel;
+            this.type = pagesIndex.getType(inputChannel);
+            this.indexChannel = indexChannel;
+        }
+
+        @Override
+        public Block getPassThroughColumn(Page page)
+        {
+            Block indexes = page.getBlock(indexChannel);
+            BlockBuilder builder = type.createBlockBuilder(null, page.getPositionCount());
+            for (int position = 0; position < page.getPositionCount(); position++) {
+                if (indexes.isNull(position)) {
+                    builder.appendNull();
+                }
+                else {
+                    // table function returns index from partition start
+                    long index = BIGINT.getLong(indexes, position);
+                    // validate index
+                    if (index < 0 || index >= endOfData[inputChannel] || index >= processedPositions) {
+                        int end = min(endOfData[inputChannel], processedPositions) - 1;
+                        String message;
+                        if (end >= 0) {
+                            message = format("Index of a pass-through row: %s out of processed portion of partition [0, %s]", index, end);
```

Comment:
The message variable is overkill. Just do `throw new TrinoException(...)` in each of the branches.